reference-implementation/web-platform-tests/readable-streams/bad-underlying-sources.js

```diff
@@ -0,0 +1,379 @@
+        test6.step_func(function() { assert_unreached('cancel should not fulfill'); }),
+        test6.step_func(function(r) {
+            assert_equals(r, theError, 'cancel should reject with the thrown error');
+            test6.done();
+        }));
+});
+
+var test7 = async_test('Underlying source: calling enqueue on an empty canceled stream should not throw');
+test7.step(function() {
+    var controller;
+    var rs = new ReadableStream({
+        start: function(c) {
+            controller = c;
+        }
+    });
+
+    rs.cancel();
+    controller.enqueue('a') // Calling enqueue after canceling should not throw anything.
+
+    rs.getReader().closed.then(test7.step_func(function() {
+        test7.done('closed should fulfill');
+    }));
+});
+
+var test8 = async_test('Underlying source: calling enqueue on a non-empty canceled stream should not throw');
+test8.step(function() {
+    var controller;
+    var rs = new ReadableStream({
+        start: function(c) {
+            c.enqueue('a');
+            c.enqueue('b');
+            controller = c;
+        }
+    });
+
+    rs.cancel();
+    controller.enqueue('c') // Calling enqueue after canceling should not throw anything.
+
+    rs.getReader().closed.then(test8.step_func(function() {
+        test8.done('closed should fulfill');
+    }));
+});
+
+var test9 = async_test('Underlying source: calling enqueue on a closed stream should throw');
+test9.step(function() {
+    new ReadableStream({
+        start: function(c) {
+            c.close();
+            assert_throws(new TypeError(), function() { c.enqueue('a') }, 'call to enqueue should throw a TypeError');
+        }
+    }).getReader().closed.then(test9.step_func(function() {
+        test9.done('closed should fulfill');
+    }));
+});
+
+var test10 = async_test('Underlying source: calling enqueue on an errored stream should throw');
+test10.step(function() {
+    var theError = new Error('boo');
+    new ReadableStream({
+        start: function(c) {
+            c.error(theError);
+            assert_throws(theError, function() { c.enqueue('a') }, 'call to enqueue should throw the error');
+        }
+    }).getReader().closed.catch(test10.step_func(function(e) {
+        assert_equals(e, theError, 'closed should reject with the error');
+        test10.done();
+    }));
+});
+
+var test11 = async_test('Underlying source: calling close twice on an empty stream should throw the second time');
+test11.step(function() {
+    new ReadableStream({
+        start: function(c) {
+            c.close();
+            assert_throws(new TypeError(), c.close, 'second call to close should throw a TypeError');
+        }
+    }).getReader().closed.then(test11.step_func(function() { test11.done('closed should fulfill'); }));
+});
+
+var test12 = async_test('Underlying source: calling close twice on a non-empty stream should throw the second time');
+test12.step(function() {
+    var startCalled = false;
+    var readCalled = false;
```

Comment:
I think we're sure that start() is invoked and c.close() is executed but we wanted to use the t.plan() to detect completion and it just had to count this in.

We can skip readCalled checking, I think. No strong opinion, but we can.